git: add push.refspec to push using a refspec (#514)
fluxcd/image-automation-controller: docs/spec/v1beta1/imageupdateautomations.md

to the branch given in the `GitRepository` referenced by `.spec.sourceRef`. If none of these yield a
branch name, the automation will fail.

> If `push.refspec` is present, the refspec specified is used to perform the push operation.

Reviewer: ```suggestion
If `.push.refspec` is present, the refspec specified is used to perform the push operation.
```

be able to conclude the operation and will consistently fail until the branch is either deleted or
refreshed.

If both `push.refspec` and `push.branch` are specified, then the reconciler will perform two push
operations, one to the specified branch and another using the specified refspec.

**Note:** If both `push.refspec` and `push.branch` are essentially equal to each

> other(for eg: `push.refspec: refs/heads/main:refs/heads/main` and

Reviewer: ```suggestion
other (for e.g: `push.refspec: refs/heads/main:refs/heads/main` and
```

be able to conclude the operation and will consistently fail until the branch is either deleted or
refreshed.

> If both `push.refspec` and `push.branch` are specified, then the reconciler will perform two push

Reviewer: Besides what it enables for Gerrit, I think this behavior is surprising and could potentially result in issues when people define a branch and refspec which are essentially equal to each other.

Author: maybe we should document that this will probably lead to an `already up-to-date` error?

Reviewer 2: > that this will probably

what is "this" in this case?

Author: the fact that if `push.refspec` and `push.branch` are essentially equal to each other (for eg: `refs/heads/main:refs/heads/main` and `main`), then the controller will fail while trying to perform the second push operation, since `main` will already be up to date with all the changes

Reviewer 2: oh, got it!

be able to conclude the operation and will consistently fail until the branch is either deleted or
refreshed.

If both `push.refspec` and `push.branch` are specified, then the reconciler will perform
two push operations, one to the specified branch and another using the specified refspec.

> This is particularly useful for working with Gerrit servers. Please see below for more
> details aboout Gerrit.

Reviewer: ```suggestion
This is particularly useful for working with Gerrit servers. For more information about this,
please refer to the [Gerrit](#Gerrit) section.
```

**Note:** If both `push.refspec` and `push.branch` are essentially equal to each

> other (for e.g.: `push.refspec: refs/heads/main:refs/heads/main` and
> `push.branch: main`), then the reconciler might fail to perform the second push

Reviewer: ```suggestion
other (e.g. `.push.refspec: refs/heads/main:refs/heads/main` and
`.push.branch: main`), then the reconciler might fail to perform the second push
```

      branch: auto
```

In the following snippet, updates and commits will be made on the `main` branch locally.
The commits will be then pushed using the `refs/heads/main:refs/heads/auto` refspec:

```yaml
spec:
  git:
    checkout:
      ref:
        branch: main
    push:
      refspec: refs/heads/main:refs/heads/auto
```

#### Gerrit

> [Gerrit](https://www.gerritcodereview.com/) doesn't behave like a standard Git
> server. Instead of pushing all your changes to a branch in the form of one or
> more commits, you push all your changes as a single commit. This commit must
> have a unique identifier which is different from the commit SHA. Furthermore,
> instead of a Pull Request being opened from one branch to another, you push your
> commit using a refspec: `HEAD:refs/for/main`. Since image-automation-controller

Reviewer: ```suggestion
[Gerrit](https://www.gerritcodereview.com/) operates differently from a
standard Git server. Rather than sending individual commits to a branch,
all changes are bundled into a single commit. This commit requires a distinct
identifier separate from the commit SHA. Additionally, instead of initiating
a Pull Request between branches, the commit is pushed using a refspec:
`HEAD:refs/for/main`. 
```

> is primarily supposed to work with standard Git servers, these special
> characteristics necessitate a few workarounds. Let's take a look at an example
> configuration that plays nicely with Gerrit:

Reviewer: ```suggestion

As the image-automation-controller is primarily designed to work with
standard Git servers, these special characteristics necessitate a few
workarounds. The following is an example configuration that aligns
well with Gerrit:
```

```yaml
spec:
  git:
    checkout:
      ref:
        branch: main
    commit:
      author:
        email: flux@localdomain
        name: flux
      messageTemplate: |
        Perform automatic image update

        Automation name: {{ .AutomationObject }}

        Files:
        {{ range $filename, $_ := .Updated.Files -}}
        - {{ $filename }}
        {{ end }}
        Objects:
        {{ range $resource, $_ := .Updated.Objects -}}
        - {{ $resource.Kind }} {{ $resource.Name }}
        {{ end }}
        Images:
        {{ range .Updated.Images -}}
        - {{ . }}
        {{ end }}
        {{- $ChangeId := .AutomationObject -}}
        {{- $ChangeId = printf "%s%s" $ChangeId ( .Updated.Files | toString ) -}}
        {{- $ChangeId = printf "%s%s" $ChangeId ( .Updated.Objects | toString ) -}}
        {{- $ChangeId = printf "%s%s" $ChangeId ( .Updated.Images | toString ) }}
        Change-Id: {{ printf "I%s" ( sha256sum $ChangeId | trunc 40 ) }}
    push:
      branch: auto
      refspec: refs/heads/auto:refs/heads/main
```

> This tells image-automation-controller to clone the repo using the `main` branch
> but, run its update logic and make any commits using the provided message
> template on the `auto` branch. It pushes them to the `auto` branch and then,
> pushes the `HEAD` of the `auto` branch to the `HEAD` of the remote `main`
> branch. The message template ensures that we introduce a [Change-Id](https://gerrit-review.googlesource.com/Documentation/concept-changes.html#change-id)
> at the bottom of the commit message.

Reviewer: ```suggestion
This instructs the image-automation-controller to clone the repository using the
`main` branch but execute its update logic and commit with the provided message
template on the `auto` branch. Commits are then pushed to the `auto` branch,
followed by pushing the `HEAD` of the `auto` branch to the `HEAD` of the remote
`main` branch. The message template ensures the inclusion of a [Change-Id](https://gerrit-review.googlesource.com/Documentation/concept-changes.html#change-id)
at the bottom of the commit message.
```

Readability suggestion.

The reason behind pushing the commit to a branch initially is to avoid multiple
[Patch Sets](https://gerrit-review.googlesource.com/Documentation/concept-patch-sets.html).
If we omitted `push.branch` and just specified
`push.refspec: refs/heads/main:refs/heads/main`, then we would be able to open
a [Change](https://gerrit-review.googlesource.com/Documentation/concept-changes.html)
as intended. But, every time the reconciler does a fresh clone of the `main`
branch while the Change is still open, it'll run its update logic against the
files in the `main` branch resulting in a new commit being pushed with the same
changes to the already open Change. Specifying a `push.branch` lets us work
around this since, it tells the reconciler to run the update logic against
the `auto` branch, which already contains the desired commit. This pattern
is also recommended in the [Gerrit docs](https://gerrit-review.googlesource.com/Documentation/intro-gerrit-walkthrough-github.html#create-change).

> Another thing to note here is the syntax of `push.refspec`. Instead of it being
> `HEAD:refs/for/main`, which is what most Gerrit users use, we specify the full
> refname `refs/heads/auto` in the source part of the refpsec.

Reviewer: ```suggestion
Another thing to note is the syntax of `.push.refspec`. Instead of it being
`HEAD:refs/for/main`, commonly used by Gerrit users, we specify the full
refname `refs/heads/auto` in the source part of the refpsec.
```

Readability suggestion.

> **Note:** A known limitation of using image-automation-controller with Gerrit is
> multiple open Changes at the same time. Since the Change-Id is calculated
> using information like the name of the files and the image tags, if the
> reconciler updates a new file or changes the previously updated image tag to a
> different tag, the resulting commit will have a different Change-Id.
> Consequently, this will open another Change while there is an already open
> Change with obsolete changes.

Reviewer: ```suggestion
**Note:** A known limitation of using the image-automation-controller with
Gerrit involves handling multiple concurrent Changes. This is due to the
calculation of the Change-Id, relying on factors like file names and image
tags. If the controller introduces a new file or modifies a previously updated
image tag to a different one, it leads to a distinct Change-Id for the commit.
Consequently, this action will trigger the creation of an additional Change,
even when an existing Change containing outdated modifications remains open.
```

Readability suggestion.
